server/services/offer.service.js

```diff
@@ -358,67 +360,73 @@ export const createOffer = async (offer) => {
 };
 
 export const acceptOffer = async (offerToAccept) => {
-  const offer = await getOffer(offerToAccept.offerId, offerToAccept.user).catch((error) => {
+  const [offer] = await getOffer(offerToAccept.offerId, offerToAccept.user).catch((error) => {
     throw new ErrorHandler(httpStatus.INTERNAL_SERVER_ERROR, 'Internal Server Error', error);
   });
 
-  if (offer.length < 1) {
+  if (!offer) {
     throw new ErrorHandler(httpStatus.PRECONDITION_FAILED, 'Invalid offer');
   }
 
-  if (offer[0].userId.toString() !== offerToAccept.user._id.toString()) {
+  if (offer.userId.toString() !== offerToAccept.user._id.toString()) {
     throw new ErrorHandler(
       httpStatus.PRECONDITION_FAILED,
       'You cannot accept offer of another user',
     );
   }
 
   if (
-    offer[0].status === OFFER_STATUS.CANCELLED ||
-    offer[0].status === OFFER_STATUS.NEGLECTED ||
-    offer[0].status === OFFER_STATUS.REJECTED
+    offer.status === OFFER_STATUS.CANCELLED ||
+    offer.status === OFFER_STATUS.NEGLECTED ||
+    offer.status === OFFER_STATUS.REJECTED
   ) {
     throw new ErrorHandler(httpStatus.PRECONDITION_FAILED, 'You cannot accept a cancelled offer');
   }
-  const propertyPrice = offer[0].propertyInfo.price;
-  const offerPrice = offer[0].totalAmountPayable;
+  const propertyPrice = offer.propertyInfo.price;
+  const offerPrice = offer.totalAmountPayable;
   const contributionReward = propertyPrice - offerPrice < 0 ? 0 : propertyPrice - offerPrice;
 
-  const expiryDate = new Date(offer[0].expires);
+  const expiryDate = new Date(offer.expires);
   if (Date.now() > expiryDate) {
     throw new ErrorHandler(httpStatus.PRECONDITION_FAILED, 'Offer has expired');
   }
 
   if (
-    offer[0].status === OFFER_STATUS.INTERESTED ||
-    offer[0].status === OFFER_STATUS.ASSIGNED ||
-    offer[0].status === OFFER_STATUS.ALLOCATED
+    offer.status === OFFER_STATUS.INTERESTED ||
+    offer.status === OFFER_STATUS.ASSIGNED ||
+    offer.status === OFFER_STATUS.ALLOCATED
   ) {
     throw new ErrorHandler(httpStatus.PRECONDITION_FAILED, 'Offer has been accepted');
   }
 
-  const paymentSchedule = generatePaymentSchedules(offer[0]);
+  const paymentSchedule = generatePaymentSchedules(offer);
 
   const nextPayment = {
     expectedAmount: paymentSchedule[0].amount,
     expiresOn: paymentSchedule[0].date,
-    offerId: offer[0]._id,
-    propertyId: offer[0].propertyId,
-    userId: offer[0].userId,
-    vendorId: offer[0].vendorId,
+    offerId: offer._id,
+    propertyId: offer.propertyId,
+    userId: offer.userId,
+    vendorId: offer.vendorId,
   };
 
-  const vendor = await getUserById(offer[0].vendorId);
+  const vendor = await getUserById(offer.vendorId);
+
+  const { isReferred, referralId } = await userIsReferredWithoutActiveReferral(offer.userId);
```

Comment:
I think we don't need the `isReferred` variable here as we can make use of `referralId` to confirm if a user has a pending referral. If `referralId` is null, then `isReferred` will be false, hence making it obsolete.

Also, we can update the function to `getPendingUserReferral` and return the `referralId` only.

```suggestion
  const referralId = await getPendingUserReferral(offer.userId);
```